File: grida-domains/domain/src/main/java/org/grida/domain/diary/DiaryService.java
package org.grida.domain.diary;

import lombok.RequiredArgsConstructor;
import org.grida.datetime.DateTimePicker;
import org.grida.domain.user.UserRepository;
import org.grida.transaction.TransactionHandler;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class DiaryService {

    private final DiaryRepository diaryRepository;
    private final UserRepository userRepository;
    private final DiaryImageRepository diaryImageRepository;
    private final DateTimePicker dateTimePicker;
    private final TransactionHandler transactionHandler;

    public long append(String userEmail, DiaryContents contents) {
        return diaryRepository.save(
                userEmail,
                contents,
                userRepository.findRoleByEmail(userEmail).getDiaryImageRefreshChance(),
                dateTimePicker.now()
        );
    }

    public long modifyContents(long diaryId, DiaryContents contents) {
        return diaryRepository.modifyContents(diaryId, contents, dateTimePicker.now());
    }

    public void delete(long diaryId) {
        transactionHandler.execute(() -> {
            diaryImageRepository.deleteAllByDiaryId(diaryId);
            diaryRepository.delete(diaryId);
        });
    }
}

File: grida-core/core-domain/src/main/kotlin/org/grida/domain/diary/DiaryService.kt
package org.grida.domain.diary

import org.grida.datetime.DateTimePicker
import org.grida.domain.base.AccessManager
import org.grida.domain.user.UserRepository
import org.grida.error.CannotAppendDiaryAtDate
import org.grida.error.CannotAppendDiaryAtFuture
import org.grida.error.GridaException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class DiaryService(
    private val diaryRepository: DiaryRepository,
    private val userRepository: UserRepository,
    private val accessManager: AccessManager,
) {

    @Transactional
    fun appendDiary(diary: Diary): Long {
        if (diary.isDateAfter(DateTimePicker.now().toLocalDate())) {
            throw GridaException(CannotAppendDiaryAtFuture)
        }

        if (diaryRepository.existsByUserIdAndTargetDate(diary.userId, diary.targetDate)) {
            throw GridaException(CannotAppendDiaryAtDate)
        }

        val targetUser = userRepository.findById(diary.userId)
        return diaryRepository.save(diary, targetUser)
    }

    fun readDiary(
        diaryId: Long,
        userId: Long,
    ): Diary {
        val diary = diaryRepository.findById(diaryId)
        accessManager.ownerOnly(diary, userId)
        return diary
    }

    fun readMonthlyDiaries(
        userId: Long,
        year: Int,
        month: Int
    ): List<Diary> {
        val range = DateTimePicker.monthOfYearRange(year, month)
        return diaryRepository.findAllByUserIdAndTargetDateBetween(userId, range)
    }

    @Transactional
    fun modify(
        diaryId: Long,
        userId: Long,
        content: String,
    ): Long {
        val diary = diaryRepository.findById(diaryId)
        accessManager.ownerOnly(diary, userId)
        diaryRepository.updateContent(diaryId, content)
        return diaryId
    }
}
